# Practicum Problems

In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

____
## Problem 1

In [2]:
# import excel
retail_df = pd.read_excel('./data/Online Retail.xlsx')

retail_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

In [3]:
# prepare data
retail_df['Description'] = retail_df['Description'].str.strip()
retail_df.dropna(axis = 0, subset=['InvoiceNo'], inplace = True)
retail_df['InvoiceNo'] = retail_df['InvoiceNo'].astype('str')
retail_df = retail_df[~retail_df['InvoiceNo'].str.contains('C')]

retail_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899          CHILDREN'S APRON DOLLY GIRL         6   
541906    581587     23254         CHILDRENS CUTLERY DOLLY GIRL         4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138         BAKING SET 9 PIECE RETROSPOT         3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[532621 rows x 8 columns]

In [4]:
# group data and one-hot encode 
Basket = (retail_df[retail_df['Country']=="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

def sum_to_boolean(x):
    if x<=0:
        return 0
    else:
        return 1
    
Basket = Basket.applymap(sum_to_boolean)

Basket.drop('POSTAGE', inplace=True, axis=1)

Basket

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                            0                           0   
536852                            0                           0   
536974                            0                           0   
537065                            0                           0   
537463                            0                           0   
...                             ...                         ...   
580986                            0                           0   
581001                            0                           0   
581171                            0                           0   
581279                            0                           0   
581587                            0                           0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                               0                                0   
536852                               0                                0   
536974                               0                                0   
537065                               0                                0   
537463                               0                                0   
...                                ...                              ...   
580986                               0                                0   
581001                               0                                0   
581171                               0                                0   
581279                               0                                0   
581587                               0                                0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                   0   
536852                                   0   
536974                                   0   
537065                                   0   
537463                                   0   
...                                    ...   
580986                                   0   
581001                                   0   
581171                                   0   
581279                                   0   
581587                                   0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                         0                            0   
536852                                         0                            0   
536974                                         0                            0   
537065                                         0                            0   
537463                                         0                            0   
...                                          ...                          ...   
580986                                         0                            0   
581001                                         0                            0   
581171                                         0                            0   
581279                                         0                            0   
581587                                         0                            0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                               0                                   0   
536852                               0                                   0   
536974                               0                                   0   
537065                               0                                   0   
537463                         

In [5]:
# extract frequent itemsets & find itemset having the largest support
Frequent_itemsets = apriori(Basket, min_support = 0.05, use_colnames = True)
Frequent_itemsets.sort_values('support', ascending = False).head(1)

support              itemsets
46  0.188776  (RABBIT NIGHT LIGHT)

In [6]:
# extract the association rules with the highest 'confidence'
Asso_Rules = association_rules(Frequent_itemsets, metric = "confidence")
Asso_Rules.sort_values('confidence',ascending = False).head(1)

antecedents  \
12  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   

                      consequents  antecedent support  consequent support  \
12  (SET/6 RED SPOTTY PAPER CUPS)            0.102041            0.137755   

    support  confidence      lift  leverage  conviction  
12  0.09949       0.975  7.077778  0.085433   34.489796

In [7]:
# extract the association rules with the highest 'Lift'
Asso_Rules = association_rules(Frequent_itemsets, metric = "lift")
Asso_Rules.sort_values('lift',ascending = False).head(1)

antecedents                     consequents  \
38  (PACK OF 6 SKULL PAPER CUPS)  (PACK OF 6 SKULL PAPER PLATES)   

    antecedent support  consequent support  support  confidence       lift  \
38            0.063776            0.056122  0.05102         0.8  14.254545   

    leverage  conviction  
38  0.047441    4.719388

__Conclusion:__

Itemset 'RABBIT NIGHT LIGHT' has the highest support with 0.1887

Highest Confidence of 0.975 is found for the following association (antecedents -> consequents): 
* SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED SPOTTY PAPER PLATES -> SET/6 RED SPOTTY PAPER CUPS

Highest Lift of 14.25 is found for the following association (antecedents -> consequents):
* PACK OF 6 SKULL PAPER CUPS -> PACK OF 6 SKULL PAPER PLATES

The rule with highest confidence is not the same as rule with highest lift. This is because confidence only takes into account the support of antecedents to find the likeliness of occurrence of consequent in the basket. While lift is the rise in probability of having consequent in the basket with the knowledge of antecedent being present.
___

## Problem 2

In [8]:
# import csv
df = pd.read_csv('./data/75000-out2-binary.csv')
df

Transaction Number  Chocolate Cake  Lemon Cake  Casino Cake  \
0                       1               0           0            0   
1                       2               0           0            0   
2                       3               0           0            0   
3                       4               0           0            0   
4                       5               0           0            0   
...                   ...             ...         ...          ...   
74995               74996               0           0            0   
74996               74997               0           0            0   
74997               74998               0           0            0   
74998               74999               0           0            0   
74999               75000               0           0            0   

       Opera Cake  Strawberry Cake  Truffle Cake  Chocolate Eclair  \
0               0                0             0                 0   
1               0                0             0                 0   
2               1                0             0                 0   
3               0                0             1                 0   
4               0                0             0                 1   
...           ...              ...           ...               ...   
74995           0                1             0                 0   
74996           0                0             0                 0   
74997           0                0             0                 0   
74998           0                0             0                 0   
74999           1                0             0                 0   

       Coffee Eclair  Vanilla Eclair  ...  Lemon Lemonade  Raspberry Lemonade  \
0                  0               0  ...               0                   0   
1                  1               0  ...               0                   0   
2                  0               0  ...               0                   0   
3                  0               0  ...               0                   0   
4                  0               0  ...               0                   0   
...              ...             ...  ...             ...                 ...   
74995              0               0  ...               1                   0   
74996              0               0  ...               0                   0   
74997              1               0  ...               0                   0   
74998              0               0  ...               0                   0   
74999              0               0  ...               0                   0   

       Orange Juice  Green Tea  Bottled Water  Hot Coffee  Chocolate Coffee  \
0                 0          0              0           0                 0   
1                 0          0              0           1                 0   
2                 1          0              0           0                 0   
3                 0          0              0           0                 0   
4                 1          0              0           0                 0   
...             ...        ...            ...         ...               ...   
74995             0          0              0           0                 0   
74996             0          0              0           0                 0   
74997             0          0              0           0                 0   
74998             1          0              0           0                 0   
74999             0          0              1           0                 0   

       Vanilla Frappuccino  Cherry Soda  Single Espresso  
0                        0            0                0  
1                        0            0                0  
2                        0            0                0  
3                        1            0                0  
4                        0            0                0  
...                    ...          ...          

In [9]:
item1_name = 'Chocolate Coffee'
item2_name = 'Chocolate Cake'

selection = df[[item1_name,
                  item2_name]]

item1_count = selection[item1_name] == 1
item2_count = selection[item2_name] == 1

f = selection.groupby([item1_count,
                       item2_count]).count()
f

Chocolate Coffee  Chocolate Cake
Chocolate Coffee Chocolate Cake                                  
False            False                      65802           65802
                 True                        2962            2962
True             False                       2933            2933
                 True                        3303            3303

In [10]:
print(f'Chocolate Coffee => Chocolate Cake : {selection["Chocolate Coffee"].corr(selection["Chocolate Cake"])}')
print(f'Chocolate Cake => Chocolate Coffee : {selection["Chocolate Cake"].corr(selection["Chocolate Coffee"])}')

Chocolate Coffee => Chocolate Cake : 0.48556649252787826
Chocolate Cake => Chocolate Coffee : 0.48556649252787837


__Conclusion:__

Chocolate Coffee and Chocolate Cake items are symmetric binary variables.

Mathemathecally correlation coefficient Φ = <br>FREQ<sub>11</sub> * FREQ<sub>00</sub> - FREQ<sub>10</sub> * FREQ<sub>01</sub> / sqrt(FREQ<sub>1x</sub> * FREQ<sub>x1</sub> * FREQ<sub>0x</sub> * FREQ<sub>x0</sub>)

where,<br>
FREQ<sub>11</sub> = Both Occur<br>
FREQ<sub>00</sub> = None Occur<br>
FREQ<sub>10</sub> = Only First one occurs<br>
FREQ<sub>01</sub> = Only Second one occurs<br>
FREQ<sub>1x</sub> = All where 1st is occuring irrespective of what second item is<br>
FREQ<sub>x1</sub> = All where 2nd is occuring irrespective of what first item is<br>
FREQ<sub>0x</sub> = All where 1st is not occuring irrespective of what second item is<br>
FREQ<sub>x0</sub> = All where 2nd is not occuring irrespective of what first item is<br>

Therefore changing the order of {Chocolate Coffee} => {Chocolate Cake} to {Chocolate Cake} => {Chocolate Coffee} makes no difference on the outcome. Correlation coefficient Φ simply measures the behavior of both feeatures and compares them.